### ISSUE TITLE
Different Single Item vs Batch Embeddings

### ISSUE THREAD

**Issue by skychwang** [ilaria-manco/muscall#5]
Hi there,

  I've trained a MusCALL model and am encoding audios with `model.encode_audio(audios)`, using the encode_audio function in [the muscall module](https://github.com/ilaria-manco/muscall/blob/main/muscall/models/muscall.py). An interesting thing that I'm finding is when I encode audios in batches, the embeddings that this produces are different than when I encode them one by one. I've double-checked as best I can, and am using `torch.no_grad()` and `model.eval()`, so I don't think that the reason behind this is that model is still passing gradients. 

  Is this correct behavior? The use case I have in mind is that I want to find the closest audio to a given audio, which involves computing the embeddings and then calculating cosine similarities. I compute the embedding to the single audio piece, and this matches the embedding of that single audio piece when I encode it again later, but this does not match the embedding of that audio piece when I encode it again as part of a batch. 

  Looks like this difference first happens specifically in `self.audio_backbone(audio)`, which corresponds to `ModifiedResNet`'s forward pass.
 
```
from muscall.models.muscall import MusCALL
from omegaconf import OmegaConf
import torch
import numpy as np
from tqdm import tqdm

AUDIO_PATHS = ['/../2.npy', '/../1.npy'] # audio paths, actual paths hidden
DEVICE = 'cuda'
MODEL_DIR = '' # model dir
path_to_model = MODEL_DIR + 'best_model.pth.tar'
muscall_config = OmegaConf.load(MODEL_DIR + 'config.yaml')

# load model
model = MusCALL(muscall_config.model_config)
checkpoint = torch.load(path_to_model)
model.load_state_dict(checkpoint["state_dict"])
model.to(DEVICE)
model.eval()

### SINGLE ENCODE ###

audio_path = AUDIO_PATHS[0]
audio_npy = np.load(audio_path)
assert audio_npy.shape[0] == 20000
audio_torch = torch.from_numpy(np.array([audio_npy])).to(DEVICE)
with torch.no_grad(): # https://discuss.pytorch.org/t/assignment-to-split-tensor-causes-memory-leak/74875
    audio_encoded = model.encode_audio(audio_torch)
    audio_encoded = audio_encoded / audio_encoded.norm(dim=-1, keepdim=True)
print(audio_encoded.shape, 'audio_encoded shape')

### BATCH ENCODE ###

for audio in tqdm(AUDIO_PATHS):
    try:
        audio_npy = np.load(audio)
        assert audio_npy.shape[0] == 20000
        audios.append(audio_npy)
    except:
        print(audio, 'error')
        continue
audios = torch.from_numpy(np.array(audios))

encoded_audios = []
with torch.no_grad(): # https://discuss.pytorch.org/t/assignment-to-split-tensor-causes-memory-leak/74875
    batch = audios.to(DEVICE)
    batch_encoded = model.encode_audio(batch)
    batch_encoded = batch_encoded / batch_encoded.norm(dim=-1, keepdim=True)
encoded_audios = batch_encoded
encoded_audios.to(DEVICE)
print(encoded_audios.shape, 'encoded_audios shape')

### SINGLE TEST 2 ###

audio_path = AUDIO_PATHS[0]
audio_npy2 = np.load(audio_path)
assert audio_npy2.shape[0] == 20000
audio_torch2 = torch.from_numpy(np.array([audio_npy2])).to(DEVICE)
with torch.no_grad(): # https://discuss.pytorch.org/t/assignment-to-split-tensor-causes-memory-leak/74875
    audio_encoded2 = model.encode_audio(audio_torch2)
    audio_encoded2 = audio_encoded2 / audio_encoded2.norm(dim=-1, keepdim=True)
print(audio_encoded2.shape, 'audio_encoded2 shape')

### SIMILARITY ###

cos = torch.nn.CosineSimilarity(dim=1, eps=1e-6)
similarities = cos(audio_encoded.to('cpu'), encoded_audios.to('cpu'))
print('top closest audios to the AUDIO:', audio_path)
top = torch.topk(similarities, 2)
print(top.indices)
print(top.values)
audio_paths = [AUDIO_PATHS[i] for i in top.indices]
for i, audio_path in enumerate(audio_paths):
    print(i, audio_path)

cos = torch.nn.CosineSimilarity(dim=1, eps=1e-6)
print(cos(audio_encoded.to('cpu'), audio_encoded2.to('cpu')))

```

Outputs

```
torch.Size([1, 512]) audio_encoded shape
100%|████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 3720.00it/s]
torch.Size([2, 512]) encoded_audios shape
torch.Size([1, 512]) audio_encoded2 shape
top closest audios to the AUDIO: ...2.npy
tensor([1, 0])
tensor([0.7100, 0.6899])
0 ...2.npy
1 ....1.npy
tensor([1.])
```

**Comment by ilaria-manco** [comment]
Hi @skychwang, thank you for opening the issue and sorry for the late response!

I haven't had time to test thoroughly, but the culprit may be [`torchaudio.transforms.AmplitudeToDB`](https://github.com/ilaria-manco/muscall/blob/main/muscall/modules/audio_backbones.py#L154). To have consistent behaviour between single and batched items, this should receive inputs of shape [`(batch, channel, freq, time)`](https://github.com/mthrok/audio/commit/1831147a035896a4587ebfa387e719862da1da1c#diff-f69839488ce70c39d8c3db8c87790ac963364faf8aefcb38ff5b026ee9dca8c7R378).

To test whether this is the issue, you could try changing [L212](https://github.com/ilaria-manco/muscall/blob/main/muscall/modules/audio_backbones.py#L212) to `spec = self.amplitude_to_db(spec.unsqueeze(1))`.

**Comment by skychwang** [comment]
Thanks @ilaria-manco, and no worries at all :) 

I think that worked! Not exactly that change, but changing L212 and L213 from

```
spec = self.amplitude_to_db(spec)
x = spec.unsqueeze(1)
```

to 

```
x = self.amplitude_to_db(spec.unsqueeze(1))
```

did the trick. 

Cosine similarity of item in batch with single item encoding previously = 0.3793, after the change it's 1.0 as expected. 

Thanks! I'll add a pull request with the change and close the issue.
